x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TrainedModel.java

```diff
@@ -30,6 +30,25 @@
      */
     InferenceResults infer(Map<String, Object> fields, InferenceConfig config, @Nullable Map<String, String> featureDecoderMap);
 
+    /**
+     * Same as {@link TrainedModel#infer(Map, InferenceConfig, Map)} but the features are already extracted.
+     */
+    InferenceResults infer(double[] features, InferenceConfig config);
+
+    /**
+     * This optimizes the model for inference.
+     *
+     * WARNING: Some models may not be serializable after being optimized
+     * @param isTopLevelModel Indicates if this model is nested under another model
+     * @param newFeatureIndexMapping New feature index mapping so that features can be extracted by a parent model
+     */
+    void optimizeForInference(boolean isTopLevelModel, Map<String, Integer> newFeatureIndexMapping);
```

Comment:
This is the most controversial part of the PR. Coincidentally, it is part of the code I hate the most. 

I opted for not returning a NEW model as these can have a fairly large foot print. If anything, this method could return a new "optimized" model type that only satisfies the `infer` methods. But this would be a new interface, set of classes (though they wouldn't have to be serializable), and possible code duplication. Also, this new "optimized model" would probably still have a large foot print as it will have to duplicate much of the data stored in the original model.

Open for discussion around this.

Reply:
Making this PR a WIP. I think the way to avoid this is by creating a `ModelView` class. It won't copy items, just contain references to the things it needs.

This mulling around on this.